### DISCUSSION TITLE
Is ssrExchange with network-only/cache-and-network useless?

### DISCUSSION THREAD

**Discussion by mikehaertl** [urql-graphql/urql#2790] [Q&A]
When you use `network-only` with `ssrExchange` you still get 2 requests: One on the server and one on the client. This happens even though we sent the data to rehydrate to the client. So I would expect there should not be another request on the client.

 This is caused (yet [again](https://github.com/FormidableLabs/urql/issues/2689)) by this change in the ssrExchange: #2198. It filters `network-only` requests and forwards them to the `fetchExchange` even if they were found in the rehydrated data.

So IMO this makes the whole `ssrExchange` pointless if you want to use `network-only`. I still don't get why `ssrExchange` should care about the cache policy. In my understanding all it should do is record responses on the server and inject (and not forward) them on the client.

Sidenote: Actually a while ago I even did a rewrite of `ssrExchange` in my attempt to fix my #2689. I had trouble to follow the current implementation as to me there's a clear separation of the exchange's duty on the client and server side. This is not fully reflected in the code and to me the exchange also messes with things it should not care about.

Let me know if you're interested in discussing this refactoring in a PR.

**Comment by JoviDeCroock** [comment]
I mean, yes, the premise of using `network-only` is to bypass caches. I don't see much use for the requestPolicy personally as it implies that you always want the freshest data which means you are conveying to `urql` that the data changes in a non-observable way (no subscriptions triggering the change coming externally or no mutations that change the data internally).

**Reply by mikehaertl** [reply]
Hmm, I tend to disagree. There may be a mix of policies: Some requests I want to cache, others not.

So again: Why should the `ssrExchange` care about cache policy *at all*? That's in the domain of `cacheExchange`. Both should be completely decoupled IMO. Right now this coupling causes quite some confusion and unexpected (and hard to fix) behavior for me.

**Reply by JoviDeCroock** [reply]
I mean, the `ssrExchange` is a type of `cacheExchange` with the difference that it's a cache conveyed between server and client. If you want stuff to be cached why use something like `network-only` and not [upgrade your requests yourself](https://github.com/FormidableLabs/urql/tree/main/exchanges/request-policy). If you are using `network-only` you will never hit your cache after rehydrating either way so I don't really see the point 😅 

would you mind giving some cases for this? I have worked on a lot of apps and haven't seen a use case for `network-only` that doesn't do a request but rehydrates the cache, we also haven't heard of this case before so a use case or reason why would be appreciated to broaden our view on the issue you are facing. Putting it a bit bluntly, you can emphasise _again_ and _yet again_ all you like but if we are talking from two different view points it will be hard to come to an agreement.

**Comment by kitten** [comment]
> So IMO this makes the whole ssrExchange pointless if you want to use network-only.

There's a bit of an important point here. It's not a good idea to use this request policy by default. I'm fact, we wouldn't recommend altering the client's default request policy in the first place and it's more there for legacy and completeness reasons, which I regret.

What are you trying to achieve by using this policy in combination with SSR is probably the more important question? At most, say, if you're in an SSG situation, I'd expect `cache-and-network` to be used once on mount or I'd recommend the `requestPolicyExchange` to be used, but I'd find it concerning to use `network-only` in an SSR scenario on the client by default as it'd always bypass the cache, even after any render

**Reply by mikehaertl** [reply]
> It's not a good idea to use this request policy by default.

Correct me if I'm wrong but `ssrExchange` looks at the policy of the current request, not the default policy, right?

https://github.com/FormidableLabs/urql/blob/30ecefe9f69a4c33849f6c9c43bf5083b35878c2/packages/core/src/exchanges/ssr.ts#L142-L144

So it shouldn't matter if it's my default policy or not - but even if it were it should be the developers decision and not being limited by the library.

> What are you trying to achieve by using this policy in combination with SSR is probably the more important question?

Do we agree, that caching policy can strongly depend on the type of entity? E.g. some almost static categories can be cached just fine whereas maybe some invoice status should not be cached at all - as it may be modified in the backend by another user and I want the client to always see his actual status without having him to do a full page reload. Still both, categories and invoice status can be used on the same page.

Now I use SSR (i.e. Nuxt 3) for the sole purpose to speed things up on the client side.

And I want to keep the load on the API at a minimum: So say the user jumps in via a bookmarked URL on his "invoices" page. The page is rendered on the server which includes a GQL request for invoice status. When the content is delivered to and rehydrated on the client side, I don't want to have yet another GQL request for that same data that was just rendered on the server some miliseconds ago.

But when the user navigates the SPA and comes back to his invoices after a minute there should be another `network-only` request for his invoice data.

I don't think that this is such a rare scenario. Actually I would go so far to say that almost any application's entities can be separated into more and less cacheable items.

> At most, say, if you're in an SSG situation ...

Ok, please bare with me, I probably miss some aspects. But could it be that the current behavior actually only makes sense in an SSG scenario where the server generated content is cached for a much longer time?

If that's the case maybe we can find a solution that makes both of us happy.

**Reply by kitten** [reply]
> some almost static categories can be cached just fine whereas maybe some invoice status should not be cached at all

The important thing here is that, even if the data needs to be up-to-date, from a client-side perspective, you're still asking the "short-term in-memory" cache and not a long-living cache for a response, so rather than using network-only, it's likely a better fit to do one of the two things I've mentioned above.

Basically, the request policy is not a signal in urql to say how "fresh" your data should be, but solely a signal to control what happens on a client-side caching control flow level, whether a result should come from the cache, only from the cache, from the cache then the network, or never from the cache.

This isn't basically really as much of a question about "caching" as it is about state management

**Reply by mikehaertl** [reply]
I have to correct myself: I was actually using `cache-and-network` policy (I've played around with many policies). But it's the same: I see the request immediately refetched on the client side. This is caused by the cache exchange which will not find the result in the cache, so the `reexecuteOperation()` will get triggered which uses `network-only`. So it has the same effect as described before.

**Reply by kitten** [reply]
Not quite, to be honest, `cache-and-network` explicitly retrieves a result from the cache and then sends a network request regardless. The `ssrExchange` is not quite a cache, it's a network replacement, which means that the flow of rehydration is:

```
cacheExchange -(miss)-> ssrExchange (hit) --> fetchExchange skipped
```

That's with `cache-and-network`. If you're to then repeat the same operation with `cache-and-network` again, it's:

```
cacheExchange -(hit)-> result but [reexecute as network-only]
cacheExchange -(skip)-> fetchExchange
```

i.e. once the result has been returned by the `ssrExchange` once, it's not responsible for that specific result anymore.

If you are to use `network-only` however, it's:


```
cacheExchange -(miss)-> ssrExchange (miss) --> fetchExchange
```

And if you were to use the `requestPolicyExchange`, it's:

```
cacheExchange -(miss)-> ssrExchange (hit) --> fetchExchange skipped
```

And this would be consistent, i.e. on subsequent requests you'd then always see `cacheExchange` hit and no extra request unless the `requestPolicyExchange` upgrades the request policy based on its TTL, and it'd be kept as `cache-first`.

If regardless of this you're seeing something else, I'd have to see what you're doing in detail, if you'd like a thorough explanation of any behaviour

**Reply by mikehaertl** [reply]
> Not quite, to be honest, cache-and-network explicitly retrieves a result from the cache and then sends a network request regardless.
> The ssrExchange is not quite a cache, it's a network replacement, which means that the flow of rehydration is:
>```
> cacheExchange -(miss)-> ssrExchange (hit) --> fetchExchange skipped
> ```
> That's with `cache-and-network`

I understand what you're saying but I'm sorry, that's not what I see. `fetchExchange` is *not* skipped with `cache-and-network`. What really happens:

```
cacheExchange --(miss and re-xecute with network-only)--> ssrExchange (miss due to network-only) -> fetchExchange
```

Have a look here:

https://github.com/FormidableLabs/urql/blob/30ecefe9f69a4c33849f6c9c43bf5083b35878c2/packages/core/src/exchanges/cache.ts#L73-L76

and then here:

https://github.com/FormidableLabs/urql/blob/30ecefe9f69a4c33849f6c9c43bf5083b35878c2/packages/core/src/exchanges/cache.ts#L151-L159

So now the operation hits `ssrExchange` and is happily forwared to the fetchExchange due to `network-only`.

**Reply by kitten** [reply]
Yes, that's operating as intended, however, `cache-and-network` will always cause a cache response and a subsequent refetch, `network-only` instructs all caches to be skipped. That's simply how it works. That's also part of the reason why we instead _always_ would keep the request policy at its default `cache-first`, unless you know that one of the behaviours above is what you need, preferably on a case by case (per-operation) basis.

Hence, depending on what you need, keep the request policy on `cache-first` and make use of tools like `requestPolicyExchange` or other selective and momentary upgrades in request policies.

That's a lot easier to handle and will always engage the cache only, unless you have a specific pre-condition in mind where you need to refetch some data. At least, in our original intention we consider this easier to handle and reason about

**Reply by mikehaertl** [reply]
Ok, I'm getting confused. My reasoning above is not correct. Lines 73-76 are not in effect if there's no cache hit (i.e. right after rehydration).

But I still see another request being fired right after the page loads . Even with `ssrExchange` and `cache-and-network`. I wonder if it has to do with #2688, because I don't see any dedup messages anymore when I use `ssrExchange`.


```
// ...
    exchanges: [
      debugExchange,
      dedupExchange,
      cacheExchange,
      ssr,
      authExchange<{ token: string }>({...}),
      fetchExchange,
    ],
// ...
  client.subscribeToDebugTarget((event) => {
    console.log('Debug Event: ', event);
  });
```

```
creating urql client for http://localhost:3001/graphql/
runtime-core.esm-bundler.js:1338 <Suspense> is an experimental feature and its API will likely change.
debug.ts:20 [Exchange debug]: Incoming operation:  {key: 3651420386, query: Proxy, variables: Proxy, kind: 'query', context: {…}}
debug.ts:20 [Exchange debug]: Completed operation:  {operation: {…}, data: {…}, extensions: undefined, error: undefined, hasNext: undefined}
debug.ts:20 [Exchange debug]: Incoming operation:  {key: 3651420386, query: Proxy, variables: Proxy, kind: 'query', context: {…}}
urqlClient.ts:108 Debug Event:  {timestamp: 1667927561362, source: 'cacheExchange', operation: {…}, type: 'cacheHit', message: 'The result was successfully retried from the cache'}
urqlClient.ts:109 cacheExchange cacheHit The result was successfully retried from the cache undefined
debug.ts:20 [Exchange debug]: Completed operation:  {operation: {…}, data: {…}, extensions: undefined, error: undefined, hasNext: undefined, …}
debug.ts:20 [Exchange debug]: Incoming operation:  {key: 3651420386, query: Proxy, variables: Proxy, kind: 'query', context: {…}}
urqlClient.ts:81 addAuthToOperation(): Applying auth token
urqlClient.ts:108 Debug Event:  {timestamp: 1667927561364, source: 'fetchExchange', type: 'fetchRequest', message: 'A fetch request is being executed.', operation: {…}, …}
urqlClient.ts:109 fetchExchange fetchRequest A fetch request is being executed. {url: 'http://localhost:3001/graphql/', fetchOptions: {…}}
urqlClient.ts:108 Debug Event:  {timestamp: 1667927561382, source: 'fetchExchange', type: 'fetchSuccess', message: 'A successful fetch response has been returned.', operation: {…}, …}
urqlClient.ts:109 fetchExchange fetchSuccess A successful fetch response has been returned. {url: 'http://localhost:3001/graphql/', fetchOptions: {…}, value: {…}}
debug.ts:20 [Exchange debug]: Completed operation:  {operation: {…}, data: {…}, error: undefined, extensions: undefined, hasNext: false}
```

**Reply by mikehaertl** [reply]
> Hence, depending on what you need, keep the request policy on cache-first and make use of tools like requestPolicyExchange or other selective and momentary upgrades in request policies.

Honestly - I'm close to giving up on urql 😢 

I probably have a completely different understanding of how things should work together. In my opinion the cache policies don't work as described as soon as you use `ssrExchange`. The solution with `requestPolicyExchange` to me seems like adding even more complexity on top - apart from that I don't even exactly understand what it does and why I need yet another component besides the cache that should already deal with cache policies.

Sorry for sounding so frustrated but I probably just had bad luck with urql. Everything looked so promising and made sense - but when I tried it it seems like I always hit the problematic spots and things didn't work as expected.
